src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,182 @@
+| r : zmod n → dihedral
+| sr : zmod n → dihedral
+
+namespace dihedral
+
+variables {n : ℕ}
+
+/--
+Multiplication of the dihedral group
+-/
+def mul : dihedral n → dihedral n → dihedral n
+| (r i) (r j) := r (i + j)
+| (r i) (sr j) := sr (j - i)
+| (sr i) (r j) := sr (i + j)
+| (sr i) (sr j) := r (j - i)
+
+/--
+The identity `1` is the rotation by `0`
+-/
+def one : dihedral n := r 0
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+def inv : dihedral n → dihedral n
+| (r i) := r (-i)
+| (sr i) := sr i
+
+/--
+The group structure on `dihedral n`
+-/
+instance : group (dihedral n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (a | a) (b | b) (c | c);
+    simp only [mul];
+    ring,
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (a | a),
+    { change r _ = r _,
+      rw zero_add },
+    { change sr _ = sr _,
+      rw sub_zero }
+  end,
+  mul_one := begin
+    rintros (a | a),
+    { change r _ = r _,
+      rw add_zero },
+    { change sr _ = sr _,
+      rw add_zero }
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (a | a),
+    { change r ((-a) + a) = r 0,
+      rw neg_add_self },
+    { change r (a - a) = r 0,
+      rw sub_self }
+  end }
+
+lemma r_mul_r (i j : zmod n) : r i * r j = r (i + j) := rfl
+lemma r_mul_sr (i j : zmod n) : r i * sr j = sr (j - i) := rfl
+lemma sr_mul_r (i j : zmod n) : sr i * r j = sr (i + j) := rfl
+lemma sr_mul_sr (i j : zmod n) : sr i * sr j = r (j - i) := rfl
+
+lemma one_def : (1 : dihedral n) = r 0 := rfl
+
+private def fintype_helper : (zmod n ⊕ zmod n) ≃ dihedral n :=
+{ inv_fun := λ i, match i with
+                 | (r j) := sum.inl j
+                 | (sr j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := r j
+                 | (sum.inr j) := sr j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/--
+If `0 < n`, then `dihedral n` is a finite group
+-/
+instance : fintype (dihedral n.succ) := fintype.of_equiv _ fintype_helper
+
+/--
+If `0 < n`, then `dihedral n` has `2n` elements.
+-/
+lemma card : fintype.card (dihedral n.succ) = 2 * n.succ :=
+begin
+  rw ←fintype.card_eq.mpr ⟨fintype_helper⟩,
+  change fintype.card (fin n.succ ⊕ fin n.succ) = 2 * n.succ,
+  rw [fintype.card_sum, fintype.card_fin, two_mul]
+end
+
+lemma r_one_pow (k : ℕ) : (r 1 : dihedral n) ^ k = r k :=
+begin
+  induction k with k IH,
+  { refl },
+  { rw [pow_succ, IH, r_mul_r],
+    congr' 1,
+    norm_cast,
+    rw nat.one_add }
+end
+
+lemma r_one_pow_n : (r (1 : zmod n))^n = 1 :=
+begin
+  cases n,
+  { rw pow_zero },
+  { rw [r_one_pow, one_def],
+    congr' 1,
+    simp }
+end
+
+lemma sr_mul_self (i : zmod n) : sr i * sr i = 1 := by rw [sr_mul_sr, sub_self, one_def]
+
+/--
+If `0 < n`, then `sr i` has order 2.
+-/
+lemma order_of_sr (i : zmod n.succ) : order_of (sr i) = 2 :=
+begin
+  rw @order_of_eq_prime _ _ _ _ _ _ nat.prime_two,
```

Comment:
This also works without the `@` as:
```suggestion
  rw order_of_eq_prime _ _,
  exact nat.prime_two,
```

Reply:
Is this the preferred way to provide arguments which are of the form `[fact (prime p)]`?